### ISSUE TITLE
Worker print output vanishes if not flushed (or if worker unexpectedly dies)

### ISSUE THREAD

**Issue by ericl** [ray-project/ray#4082]
```
import sys
import ray

@ray.remote
class A:
    def __init__(self):
        print("print 1")
        sys.exit(1)

@ray.remote
class B:
    def __init__(self):
        pass
    def f(self):
        print("print 2")
        sys.exit(1)

@ray.remote
def f():
    print("print 3")
    sys.exit(1)


ray.init()
A.remote()
try:
    ray.get(B.remote().f.remote())
except:
    pass
try:
    ray.get(f.remote())
except:
    pass

```

This won't print anything at all, and used to print all of 1, 2, 3.

**Comment by ericl** [comment]
@robertnishihara

**Comment by ericl** [comment]
In RLlib, this can manifest itself in mysterious hangs (somehow exceptions are swallowed as well on certain types of errors).

**Comment by robertnishihara** [comment]
I assume what's going on here is that it's being buffered and will be flushed eventually if more print statements happen on the same worker.

One thing we could potentially do is just call `sys.stdout.flush()` and `sys.stderr.flush()` on a timer in the workers.

**Comment by robertnishihara** [comment]
@ericl in your example, does the `"HI"` get printed?

**Comment by ericl** [comment]
I think there's some way to enable line buffering right?

@robertnishihara nope, you don't get "HI". I think there's also edge cases where you don't get output when the worker is unexpected killed.

**Comment by robertnishihara** [comment]
Ok, I just ran it and I saw `"HI"` (on MacOS) but I also had to do `import sys` to make it work.

As for line buffering, I tried that in

https://github.com/ray-project/ray/blob/b78d77257b515ff9130f72486e0823b91e82523c/python/ray/node.py#L227-L228

to replace `buffering=1` with `buffering=0`, but got an error

```
~/ray_scratch/ray/python/ray/tempfile_services.py in new_log_files(name, redirect_output)
    176         suffix=".err", prefix=name, directory_name=logs_dir)
    177     # Line-buffer the output (mode 1)
--> 178     log_stdout_file = open(log_stdout, "a", buffering=0)
    179     log_stderr_file = open(log_stderr, "a", buffering=0)
    180     return log_stdout_file, log_stderr_file

ValueError: can't have unbuffered text I/O
```

Replacing the `'a'` with `'ab'` gets rid of that error, but for whatever reason the print statements can still get buffered (`'wb'` behaves similarly).

It might have something to do with the way we use `dup2` in https://github.com/ray-project/ray/blob/b78d77257b515ff9130f72486e0823b91e82523c/python/ray/worker.py#L1863-L1864

@suquark any ideas about this?

**Comment by robertnishihara** [comment]
@ericl can you elaborate on how this can lead to hanging in RLlib? Is RLlib parsing stdout/stderr?

**Comment by ericl** [comment]
I'm not exactly sure what was going on. The behaviour I saw was:

Some piece of code throws an exception in RLlib. However, you don't see anything at all on the driver, and it just appears to hang. Also, the log monitor is using 100% cpu.

It's kind of weird, because I can't reproduce this with a simple example like the one above.

**Comment by ericl** [comment]
Here's a reproduction:
```
import ray
from ray.tune import run_experiments
from ray.rllib.models import Model, ModelCatalog

class CustomModel(Model):
    def _build_layers_v2(self, input_dict, num_outputs, options):
        import sys
        sys.exit(1)

ray.init()
ModelCatalog.register_custom_model("custom", CustomModel)
run_experiments({
    "custom_loss": {
        "run": "PG",
        "env": "CartPole-v0",
        "config": {
            "model": {
                "custom_model": "custom",
            },
        },
    },
})
```

Note that raising an exception is fine, however exit(1) causes a hang. In the real case the exit was caused by TF segfaulting I think.

**Comment by robertnishihara** [comment]
Ok, there are two issues here.

1. For the workload in the first comment, the prints actually do happen if the script stays alive, but if the driver exits before the prints get streamed from the log monitor to the driver then they don't get printed.

2. For the Tune workload in https://github.com/ray-project/ray/issues/4082#issuecomment-464623602, the driver was blocked in `ray.wait` which didn't release the GIL and so the logs weren't getting printed. Fixed in #4190.

**Comment by javierabosch** [comment]
I believe I am having these same issues occurring in ray 1.1.0

After a worker errors out, and I attempt to run the job again, I cannot neither print to the console or log warnings using log.warning
